### PR TITLE
feat: Table support `virtual`

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 👀 Preview

https://preview-44349-ant-design.surge.sh/components/table-cn#components-table-demo-virtual-list

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  🔥 Table support `virtual` props.        |
| 🇨🇳 Chinese |    🔥 Table 支持 `virtual` 属性开启虚拟滚动。       |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at e77224e</samp>

This pull request adds virtual scrolling support to the `Table` component using the `RcVirtualTable` component from the `rc-table` library. It also updates the documentation, the demo, and the styles for the `Table` component to reflect this new feature. Additionally, it renames the `RcTable.tsx` file to `index.tsx` and demonstrates a tooltip bug with virtual scrolling.

### <samp>🤖 Generated by Copilot at e77224e</samp>

*  Add virtual scrolling feature to table component ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R112),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R147),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R521),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616L535-R571),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-145c0de36d4564be82db77cd2f591265a596b56a4a92d0701e1f75758ffcb9deR1-R12),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-3338315f48e7215271d0f914a07795d7a7f6e18409a207cb24b864e3a0315181L2-R2),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a9612177ff7fc2e50de3a66777bcee3c89711d02a709c61366efbc3b838a8e6cR20),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a9612177ff7fc2e50de3a66777bcee3c89711d02a709c61366efbc3b838a8e6cR276),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a9612177ff7fc2e50de3a66777bcee3c89711d02a709c61366efbc3b838a8e6cR518),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-e079a40882207691b3e8205df7d39d92e7730ca90d195b3a5668b5bb0ff4901eR18),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-7c7346bcc5b9d2329966a268db0784b15a13add22938c2fe06c9efead085b3fdR1-R65),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L150-R150))
  - Import `RcVirtualTable` component from `RcTable` subdirectory and use it conditionally based on `virtual` prop in `InternalTable.tsx` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R23),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616L535-R571),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-145c0de36d4564be82db77cd2f591265a596b56a4a92d0701e1f75758ffcb9deR1-R12),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-3338315f48e7215271d0f914a07795d7a7f6e18409a207cb24b864e3a0315181L2-R2))
  - Add `virtual` prop to `TableProps` interface and pass it to `InternalTable` component in `InternalTable.tsx` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R112),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R147))
  - Use `useToken` hook to get theme token object and calculate `listItemHeight` based on table size in `InternalTable.tsx` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616R521),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a5896d4aa0af160e25dae36aec024d58f1f1769814b1777acedf581c53140616L535-R571))
  - Add `renderTimes` prop to `RcVirtualTable` component and use custom `shouldUpdate` function to trigger update when data or columns change in `VirtualTable.tsx` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-145c0de36d4564be82db77cd2f591265a596b56a4a92d0701e1f75758ffcb9deR1-R12))
  - Import `genVirtualStyle` function from `virtual.ts` and call it in `genTableStyle` function in `index.ts` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a9612177ff7fc2e50de3a66777bcee3c89711d02a709c61366efbc3b838a8e6cR20),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a9612177ff7fc2e50de3a66777bcee3c89711d02a709c61366efbc3b838a8e6cR518))
  - Add table cell selector to `genSizeStyle` function in `size.ts` and `index.ts` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-a9612177ff7fc2e50de3a66777bcee3c89711d02a709c61366efbc3b838a8e6cR276),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-e079a40882207691b3e8205df7d39d92e7730ca90d195b3a5668b5bb0ff4901eR18))
  - Export `genVirtualStyle` function that generates CSS object for virtual table styles in `virtual.ts` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-7c7346bcc5b9d2329966a268db0784b15a13add22938c2fe06c9efead085b3fdR1-R65))
  - Update dependency version of `rc-table` library in `package.json` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L150-R150))
* Update documentation and demo for virtual list in table component ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-f16bad47b640e7ccfb261714cde00e85634ec34e28a85d787cc95454467b31ddL3-R7),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-037ef5e6ac55b882162ae0ef7858ce830eab52dd172f650b2388aafffd5a0d34L1-R151),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-8d3a4194d516aa3322b22843bc2a379cdb0ad134c5e22899e6468d16c85d85c9L100-R100),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-1d555de1f30d5e46ff95e81b3d10b27baf15321e455557e7d3a55ca09ff61209L101-R101))
  - Change description of virtual list demo and add version attribute to code block in `virtual-list.md` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-f16bad47b640e7ccfb261714cde00e85634ec34e28a85d787cc95454467b31ddL3-R7))
  - Replace code for virtual list demo and use built-in `virtual` prop and `InternalTable` component in `virtual-list.tsx` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-037ef5e6ac55b882162ae0ef7858ce830eab52dd172f650b2388aafffd5a0d34L1-R151))
  - Add version attribute to virtual list demo code block in `index.en-US.md` and `index.zh-CN.md` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-8d3a4194d516aa3322b22843bc2a379cdb0ad134c5e22899e6468d16c85d85c9L100-R100),[link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-1d555de1f30d5e46ff95e81b3d10b27baf15321e455557e7d3a55ca09ff61209L101-R101))
* Modify basic demo in tooltip component to demonstrate a bug with virtual table ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-5ecb071295e03315d5d4b6e912bb0072e0fbad91283da799b97559b516ef7bfeL5-R5))
  - Add long title, bottom placement, and open prop to tooltip component in `basic.tsx` ([link](https://github.com/ant-design/ant-design/pull/44349/files?diff=unified&w=0#diff-5ecb071295e03315d5d4b6e912bb0072e0fbad91283da799b97559b516ef7bfeL5-R5))

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -147,7 +147,7 @@
     "rc-slider": "~10.1.0",
     "rc-steps": "~6.0.1",
     "rc-switch": "~4.1.0",
-    "rc-table": "~7.32.1",
+    "rc-table": "~7.33.1",
```
Comment (yoyo837) [inline package.json:150]: `package.json` 的 `react-window` 和 `@types/react-window` 是不是可以干掉了

**Comment by yoyo837** [comment]
请教为啥 `RCTable` 没有直接支持 `virtual` 而是多一个 `VirtualTable` 导出的形式呢?🤔️

**Comment by zombieJ** [comment]
> 请教为啥 `RcTable` 没有直接支持 `virtual` 而是多一个 `VirtualTable` 导出的形式呢?🤔️

因为本质上还是通过 `components` 接口自定义了 `body` 部分，并不是直接在 Table 上改的。

**Comment by yoyo837** [comment]
> > 请教为啥 `RcTable` 没有直接支持 `virtual` 而是多一个 `VirtualTable` 导出的形式呢?🤔️
> 
> 因为本质上还是通过 `components` 接口自定义了 `body` 部分，并不是直接在 Table 上改的。

那既然是分开的, 能做到没有 `virtual` 时 bundle 小一点吗(少导入东西)?

**Comment by zombieJ** [comment]
> 那既然是分开的, 能做到没有 `virtual` 时 bundle 小一点吗(少导入东西)?

那 API 上就会比较麻烦，index 里变成 `export { Table, VirtualTable }`。不能 `<Table virtual />`。理论上来说，antd 是上层封装，它通过属性透出，降低开发者的使用成本。如果像 rc-table 里一样使用就会比较裸。是个权衡问题。

File: components/table/InternalTable.tsx [hunk 2]
```diff
@@ -532,11 +538,38 @@ const InternalTable = <RecordType extends AnyObject = AnyObject>(
     <DefaultRenderEmpty componentName="Table" />
   );
 
+  // ========================== Render ==========================
+  const TableComponent = virtual ? RcVirtualTable : RcTable;
```
Comment (kiner-tang) [inline components/table/InternalTable.tsx:542]: 这样处理，是不是打包的事后会同时打包 `RcVirtualTable`  `RcTable`,导致包太大，tree shaking 好像也摇不掉吧。感觉是不是两个单独使用就好。需要完整功能的就用 RcTable,只是普通展示的简单表格就用 RcVirtualTable 呢？

**Review by li-jia-nan — APPROVED** [review]

File: components/table/InternalTable.tsx (hunk 2 above)
Comment (zombieJ) [reply components/table/InternalTable.tsx:542]: 另外一点就是虽然是两个组件，VirtualTable 其实不是重写了一个 Table，而是只有 `component: body` 部分

**Review by yoyo837 — APPROVED** [review]

**Review by kiner-tang — APPROVED** [review]

File: components/table/InternalTable.tsx (hunk 2 above)
Comment (kiner-tang) [reply components/table/InternalTable.tsx:542]: 明白了

**Comment by x8845723** [comment]
virtual再加上表格的行合并，样式就错乱了，有更新计划吗？现在只能css来handle

**Pull request merged**